subsys/usb/device_next/class/usbd_cdc_acm.c

```diff
@@ -546,6 +550,10 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
 		return;
 	}
 
+	if (atomic_test_and_set_bit(&data->state, CDC_ACM_TX_EP_BUSY)) {
```

Comment:
Shouldn't this be moved below the `if (buf == NULL)` check? Isn't this the deadlock that @SynchronicIT is referring to?

Reply:
I think what @SynchronicIT meant applies to both 
https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/subsys/usb/device_next/class/usbd_cdc_acm.c#L557-L561 

and 

https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/subsys/usb/device_next/class/usbd_cdc_acm.c#L566-L570

where in both cases, the TX_EP_BUSY status will be stuck at busy. To resolve this, we could add release for the both if. IMO, putting the `atomic_test_and_set_bit()` any further might not be what we wanted, as we would not want to allocate buffer before we know that the endpoint is ready for transmission. It looks like it's "cheaper" to set the bit first, and then clear it when either of the failure case occurs. 

@tmon-nordic , should we make this ticket as duplicate like what @SynchronicIT suggested, since his PR is created earlier? I am still in favor of this solution, but I am not sure whether that is the correct thing to do in Zephyr community? 

Thank you.

Reply:
If you add the clears then how does this PR differ from @SynchronicIT?

Reply:
Comparing the 2 PR, What I can see are these: 
1. [Similar] The flag name, and the places to set and clear the BUSY flag are the same. (after adding clears accordingly)
2. [Different] When RX complete event occurs, @SynchronicIT PR queues an additional `tx_fifo_work` directly from `usbd_cdc_acm_request()` if tx_fifo is not empty (on top of the existing `irq_cb_work`), while for my PR, we are doing the equal handling at `cdc_acm_irq_cb_handler()`. 
3. [Different] @SynchronicIT PR is using `tx_fifo.altered` to indicate new transfer, and tx_fifo empty status as indication of new transaction requested. Which my PR use only tx_fifo empty status. Although that @SynchronicIT  PR will spawn 2 work on TX complete interrupt (`irq_cb_work` + `tx_fifo_work`), the new handling in this PR itself should resolve the conflict by deferring the older one, hence should not be an issue. 

I think I somehow missed out point (2), hence thinking it would not resolve the possible cases where data written withheld in the ringbuffer issue. 
So for now, I do think that #78160, although it is not totally as what I would love the change to be, I think it resolves the issues I am trying tor resolve.

Reply:
@tmon-nordic, if I understands correctly, since it seems that #78160 is already capable of fixing all the issues we foresee, this is where I add "duplicate of #78160" comment, and close this PR right?